### PR TITLE
Space charge tests: improve backward-mode test ; add forward-mode test

### PR DESCRIPTION
**Pull request by RemiLehe**
## Description

This PR improves the space-charge tests:

- It now tests the accuracy of the gradient calculation, for the backward-mode differentiation.
- In addition, it adds a similar test for forward-mode differentiation.

## Motivation and Context

Since Cheetah is built with `pytorch`, it supports backward-differentiation, but also forward differentiation (see [this page](https://pytorch.org/tutorials/intermediate/forward_ad_usage.html)). This could be useful since the scaling of forward differentiation is different from that of backward differentiation. 

In addition, previous tests of backward-differentiation mode were simply checking that the differentiation ran without error, but they did not check the accuracy of the computed gradient. In this PR, we now test the accuracy of the gradient, by reusing the parameters of https://accelconf.web.cern.ch/hb2023/papers/thbp44.pdf, and by using an analytical formula to find the expected gradient.

## Types of changes

New test

## Checklist

- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (_required for a bug fix or a new feature_).
- [ ] I have updated the documentation accordingly.
  
  
- [ ] I have reformatted the code and checked that formatting passes (**required**).
- [ ] I have have fixed all issues found by `flake8` (**required**).
- [x] I have ensured that all `pytest` tests pass (**required**).
- [ ] I have run `pytest` on a machine with a CUDA GPU and made sure all tests pass (**required**).
- [ ] I have checked that the documentation builds (**required**).

### CODE REVIEW

File: tests/test_space_charge_kick.py [hunk 1]
```diff
@@ -200,15 +201,42 @@ def test_incoming_beam_not_modified():
...
 def test_gradient():
     """
-    Tests that the gradient of the track method is computed withouth throwing an error.
+    Tests that the gradient of the track method is computed accurately.
     """
-    incoming_beam = cheetah.ParticleBeam.from_parameters(
-        num_particles=torch.tensor(10_000),
-        sigma_px=torch.tensor(2e-7),
-        sigma_py=torch.tensor(2e-7),
+
```
Comment (RemiLehe) [inline tests/test_space_charge_kick.py:206]: This is reusing the same code as in the test `test_cold_uniform_beam_expansion` above.

**Comment by cr-xu** [comment]
Thanks for providing a forward AD test. I think similar tests could also be added to other parts, but we can do that in a separate PR.

File: tests/test_space_charge_kick.py [hunk 2]
```diff
@@ -12,7 +13,7 @@ def test_cold_uniform_beam_expansion():
     travelling through a drift section with space_charge. (cf ImpactX test:
     https://impactx.readthedocs.io/en/latest/usage/examples/cfchannel/README.html#constant-focusing-channel-with-space-charge)
```
Comment (RemiLehe) [inline tests/test_space_charge_kick.py:14]:
```suggestion
    https://impactx.readthedocs.io/en/latest/usage/examples/expanding_beam/README.html
```

**Comment by cr-xu** [comment]
Hi @RemiLehe, could you also update the CHANGELOG.md? Somehow I still couldn't directly push to your forked repo. Other than that the PR looks good to go.

**Comment by RemiLehe** [comment]
Sounds good. Done.

**Review by cr-xu — APPROVED** [review]
